File: ruby/dest_city.rb
paths = [['London', 'New York'], ['New York', 'Lima'], ['Lima', 'Sao Paulo']]
paths2 = [%w[B C], %w[D B], %w[C A]]
paths3 = [%w[A Z]]
paths4 = [%w[pYyNGfBYbm wxAscRuzOl], %w[kzwEQHfwce pYyNGfBYbm]]
paths5 = [['jMgaf WaWA', 'iinynVdmBz'], [' QCrEFBcAw', 'wRPRHznLWS'], %w[iinynVdmBz OoLjlLFzjz],
          ['OoLjlLFzjz', ' QCrEFBcAw'], ['IhxjNbDeXk', 'jMgaf WaWA'], ['jmuAYy vgz', 'IhxjNbDeXk']]

def dest_city(paths)
  paths = paths.flatten

  destination_data(paths).each do |place, _count|
    return place if destination_data(paths)[place] == 1 && paths.index(place).odd?
  end
end

def destination_data(paths, destination = {})
  paths.each do |path|
    if destination[path]
      destination[path] += 1
    else
      destination[path] = 1
    end
  end

  destination
end

p dest_city(paths)
p dest_city(paths2)
p dest_city(paths3)
p dest_city(paths4)
p dest_city(paths5)

File: python/destCity.py
paths = [['London', 'New York'], ['New York', 'Lima'], ['Lima', 'Sao Paulo']]
paths2 = [['B', 'C'], ['D', 'B'], ['C', 'A']]
paths3 = [['A', 'Z']]
paths4 = [['pYyNGfBYbm', 'wxAscRuzOl'], ['kzwEQHfwce', 'pYyNGfBYbm']]
paths5 = [['jMgaf WaWA', 'iinynVdmBz'], [' QCrEFBcAw', 'wRPRHznLWS'], ['iinynVdmBz', 'OoLjlLFzjz'],
          ['OoLjlLFzjz', ' QCrEFBcAw'], ['IhxjNbDeXk', 'jMgaf WaWA'], ['jmuAYy vgz', 'IhxjNbDeXk']]

import itertools

# This needs to be revisted and worked on!

def destCity(paths):
  cities = list(itertools.chain(*paths))

  for place in destinationData(cities).keys():
    import pdb; pdb.set_trace()
    if destinationData(cities)[place] % 2 == 1 and cities.index(place) % 2 == 1:
      return place

def destinationData(paths, destination = {}):
  for path in paths:
    if path in destination:
      destination[path] += 1
    else:
      destination[path] = 1

  return destination


print(destCity(paths))
print(destCity(paths2))
print(destCity(paths3))
print(destCity(paths4))
print(destCity(paths5))
